### ISSUE TITLE
Remove Animations Metadata from Decorators and Place it in CSS

### ISSUE THREAD

**Issue by jfftck** [angular/angular#24327]
Move the animation metadata to CSS files and allow importing from common animation CSS, as the current animations in the decorator is very limited on reuse and makes the components very messy.

The current @angular/animations module requires about 3 MB and it is very static, the useAnimation metadata is very limited on how it can be used that it is a joke. But, since CSS is static, it would be the better candidate for hosting this information.

Then support for CSS variables and injecting data into these variables would remove the static nature and allow much cleaner code for components and the templates.

Added bonus is animations can be attached to the elements without an HTML attribute.

I think the biggest issue that I have with the current animations is that they are no better than code full of magic numbers/variables and that is seen as a bad practice. 

## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Performance issue
[x] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

**Comment by jfftck** [comment]
Currently Angular does animations like this:

```javascript
...
animations: [trigger('buttonAni', [
  transition('* <=> *', [
    query(':leave', [
      style({transform: 'scaleX(1)', opacity: 1}),
      animate(
        '125ms cubic-bezier(0.4, 0.0, 0.2, 1)',
        style({transform: 'scaleX(0)', opacity: 0})
      )
    ], { optional: true }),
    query(':enter', [
      style({transform: 'scaleY(0)', opacity: 0}),
      animate(
        '500ms cubic-bezier(0.4, 0.0, 0.2, 1)',
        style({transform: 'scaleY(1)', opacity: 1})
      )
    ],
    { optional: true })
  ])
])]
...
```

```html
<div class="fake-button" [@buttonAni]>
```

But I feel that moving the metadata to CSS would clean up the code:
**button-animation.css**
```css
:root {
  --enter-timing: 500ms;
  --leave-timing: 125ms;
}

@-ng-animate(button) {
  @-ng-leave {
    @-ng-timing: var(--leave-timing);

    0% {
      transform: scaleX(1);
      opacity: 1;
    }

    100% {
      transform: scaleX(0);
      opacity: 0;
    }
  }

  @-ng-enter {
    @-ng-timing: var(--enter-timing);

    -ng-start {
      transform: scaleY(0);
      opacity: 0;
    }

    -ng-end {
      transform: scaleY(1);
      opacity: 1;
    }
  }
}
```
**fake-button.component.css**
```css
@-ng-import(button-animation.css) {
  --leave-timing: 250ms;
  --enter-timing: {{ enter-timing }};
}

.fake-button {
  -ng-animate: button;
}
```

That would allow attaching the animation to any part of a component and allow reusable animations.

Also, notice that there is custom property injection in the component CSS to the animation CSS and a component variable extraction.

_Note_
This syntax should have the following rules:
* Attaching would only be available for component CSS
* Attaching more than one animation should merge them with the priority of properties going from first to last to match standard CSS rules (box-shadow, text-shadow, background...)

This is just a simple example and does not cover all of the metadata properties that the current animation framework has, but they should follow the same kind of syntax.

**Comment by levgaas** [comment]
@jfftck thanks for your feedback.

**_Some thoughts on top of this:_**
Animation metadata is not as static as you might think =)

**Firstly**, the size of the package is not mostly caused by parsing metadata. 
But by picking proper driver, renderer, engine, etc. And transforming (delegating) them to proper implementation supported by browser.
Also handling `parallel, queued, deferred, sequenced, nested` players.

Regarding `static`:
Have a look at how to make it a little bit more dynamic:
- build on the fly and have granular control over player (https://angular.io/api/animations/AnimationBuilder)
- interpolated params just right from your component / js (https://angular.io/api/animations/animation)
- dynamic states
- dynamic trigger name

Think of `@angular/animations` more like as high level instructions (builder) of Web Animations API
In a nutshell, with that builder you could possibly construct brick by brick whatever you want which is dynamic over time or conditions.

**Comment by jfftck** [comment]
@levgaas Why not build this as a more functional set of tools, I would like to use the size information that `style({height: '*'})` on other elements to be able to have tab pages that would resize and fade between the two, but for that I would also need the `":enter"` and `":leave"` events as well.

That is why I was thinking that creating reactive CSS would kill two birds with one stone as it would render the animation trigger `[@<name>]`, `ngClass`, and `ngStyle` unnecessary and would clean up the template.

Also, why not just use a polyfill for Web Animations API for any environment that does not support it? I know that a polyfill is already being used and then this code on top of that.

I just feel that if I need to include code that it should be doing more than what the current animations module is doing.

I tried to use `useAnimation` but it can't be used directly under the trigger. I want to have a complete animation package that I could apply to components.

Why does the Angular team want a template that has so much markup that the content is hard to find?

Finally, why is the animations defined outside of the component class and not a function or object that is injected in `ngInit`?

**Comment by levgaas** [comment]
@jfftck 

Animations metadata are placed in decorator because they are transformed into instructions and registered at the moment of creating a component.

And those `trigger, animation, query, etc.` are called only once to build an AST that's used later by the `AnimationEngine`.

**Regarding placing it in html**:
When you include `BrowserAnimationsModule` or `NoopAnimationsModule`, included module provides a decorated `AnimationRenderer` instead of default one.
Which knows how to utilize existing power of Angular (listen to bindings, outputs, etc.) and reflects those changes to animations.

One additional benefit is that your html is parsed at the compile time. And can show compile time errors. Thus your favorite IDE highlights them.

`One thought:` is that the code you provided takes more lines of code than it's JS analogue. 
Plus from what I see there should be additional parser to parse CSS and transform them to instructions. I don't think it will make the package smaller =)

Regarding `useAnimation` and `trigger` functions:
You can use something like:
```
 trigger('shrinkOut', [
    transition('* => leave', [
      style({ height: '*' }),
      animate(250, style({ height: '{{ height }}px' }))
    ])
  ])
```

And in you html:
```
[@shrinkOut]="{ value: animationState, params: { height: 24 } }"
```

If you are still lost it would be a pleasure of mine to make your `animations` a little bit more reusable =)
